### DISCUSSION TITLE
Alternative to downcast DocumentSnapshot's generic

### DISCUSSION THREAD

**Discussion by stargazing-dino** [firebase/flutterfire#6179] [Q&A] [answered]
Hello !

Dart does not allow for downcasting a class's generic type. This makes things like this not possible:

```dart
@freezed
class Post with _$Post {
  const factory Post.textPost(/** ... */) = TextPost;
  const factory Post.pollPost(/** ... */) = PollPost;

  factory Post.fromJson(Map<String, dynamic> json) => _$PostFromJson(json);
}

final postsRef = _instance.collection('posts').withConverter<Post>(
      fromFirestore: (snapshot, _) => Post.fromJson(snapshot.data()!),
      toFirestore: (post, _) => post.toJson(),
    );
```

```dart
return StreamBuilder(
  stream: _postsStream,
  builder: (context, snapshot) {
    // ...

    final docs = snapshot.data.docs;

    return ListView(
      children: docs.map(
        (doc) {
          final data = doc.data();

          return data.when(
            textPost: (_textData) {
              return TextCard(
                // Fails
                doc: doc as QueryDocumentSnapshot<TextPost>
              );
            },
            pollPost: (_pollData) {
              return PollCard(
                // Fails
                doc: doc as QueryDocumentSnapshot<PollPost>
              );
            },
          );
        },
      ),
    );
  },
);
```

The error thrown is:

```console
type '_WithConverterQueryDocumentSnapshot<Post>' is not a subtype of type 'QueryDocumentSnapshot<TextPost>' in type cast
```

That's fine if we can't do the conversion but it'd be great if there was a a way to somehow change that data type in another way. Currently `.data()` can't be overriden like a mutable field can so something like this will fail:

```dart
final TextDoc = doc..data = _textData;
```

Another option would be to just recreate the class but QueryDocumentSnapshot and its parents have no public constructors.

The only other option I can think of is to create a wrapper class around DocumentSnapshot which has data as a field that I can `copyWith` to anything. I guess that would also help me separate out `firestore` stuff from my code but I've never considered that separation of layers a priority.

Any advice?

cc @rrousselGit

**Comment by rrousselGit** [comment, marked as answer]
My recommendation would be to instead do:

```dart
PollCard(
  onChange: (newPost) => doc.set(newPost),
  post: doc.data(),
);
```

**Reply by stargazing-dino** [reply]
I tried that solution but after some thought I decided to just give up and go with the wrapper class idea. No particular reason -- I guess it just looked more robust

Anyways here's the solution I decided on. [Gist of full document implementation here](https://gist.github.com/Nolence/8983a46b839c76a7770a5494f8c5af9f)

```dart
@freezed
class Document<T extends Object?> with _$Document<T> {
  const Document._();
  const factory Document({
    required DocumentReference<T> reference,
    required T data,
  }) = _Document<T>;

  // ...

  Document<R> cast<R>({
    required R data,
    required FromFirestore<R> fromFirestore,
    required ToFirestore<R> toFirestore,
  }) {
    return Document<R>(
      data: data,
      reference: reference.withConverter<R>(
        fromFirestore: fromFirestore,
        toFirestore: toFirestore,
      ),
    );
  }
}
```

Always appreciate your help remi !
